File: python/algo/202606/algo3747.py
class Solution:
    def countDistinct(self, n: int) -> int:
        s = str(n)
        # 9 + 9^2 + ... + 9^(len(s)-1)
        ans = ((9 ** len(s)) - 9) // 8 # 小于 len(s) 位数不包含 0 的数量
        
        for i, ch in enumerate(s):
            d = ord(ch) - ord('0')
            if d == 0: break # 包含 0，后续都不计入
            
            if (i == len(s) - 1): ans += d
            else: ans += (d-1) * (9 ** (len(s)-i-1))
        return ans

File: cplus/leetcode/202606/algo3747.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    long long countDistinct(long long n) {
        // 组合数学 自行解答
        string s = to_string(n);
        long long ans = (pow(9, s.size()) -9 ) / 8;

        for (int i=0; i<s.length(); i++) {
            auto d = s[i] - '0';
            if (d == 0) break;
            if (i == s.length() - 1) ans += d;
            else ans += (d-1) * pow(9, s.length()-i-1);
        }

        return ans;
    }
};